packages/@aws-cdk/cdk/lib/resource.ts

```diff
@@ -1,16 +1,59 @@
 import { Construct, IConstruct } from './construct';
+import { PhysicalName } from './physical-name';
+import { PhysicalNameGenerator } from "./physical-name-generator";
 
 /**
  * Interface for the Resource construct.
  */
-// tslint:disable-next-line:no-empty-interface
 export interface IResource extends IConstruct {
+  readonly physicalName: PhysicalName;
 
+  /**
+   * Returns the {@link PhysicalNameGenerator} for this Resource,
+   * which will be used to generate a name for it if one was not provided explicitly.
+   * The implementation in {@link Resource} returns a default generator;
+   * subclasses with particular requirements for physical names are expected to override this method,
+   * possibly by calling `super` and changing the resulting generator,
+   * to fit their needs.
+   */
+  readonly physicalNameGenerator: PhysicalNameGenerator;
```

Comment:
Still not sure where this is really needed. Is it just to support lowercasing the s3 bucket? If that's the only case, let's just make the default generator generate lowercase letter and introduce this only if it's really needed in the future. Adding APIs is easier than removing them and this also pollutes all of resources public API.

Reply:
Changed. `PhysicalNameGenerator` is now package-private, and always generates lowercase names.